Fix AWS eventual consistency of AttachDisk

jsafrane opened the pull request:
AWS eventual consistency can go back in time. It can return that a volume is detached and then that it is back attached without Kubernetes doing anything to the volume.

When this happens during attachment of the same volume to the same node, but with a different 
device name, retry `DescribeVolumes` a few times before reporting an error. 10 retries should be enough to get a consistent result.

/kind bug

```release-note
Fixed "requested device X but found Y" attach error on AWS.
```

/sig storage
/king important-longterm
@wongma7 @gnufied, PTAL

jsafrane committed:
Fix AWS eventual consistency of AttachDisk

AWS eventual consistency can go back in time. It can return that a volume
is detached and then that it is attached.

When this happens during attachment of the same volume to the same node,
but with a different device name, retry DescribeVolumes a few times before
reporting an error. 10 retries should be enough to get a consistent result.

jsafrane commented:
/retest

wongma7 commented:
/lgtm

gnufied commented:
Just putting on hold for a bit while I also have a chance to review.

/hold

gnufied commented on staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:
```diff
@@ -2177,11 +2176,29 @@ func (d *awsDisk) waitForAttachmentStatus(status string) (*ec2.VolumeAttachment,
 		}
+		if attachment != nil && expectedDevice != "" {
+			device := aws.StringValue(attachment.Device)
+			if device != "" && device != expectedDevice {
```
Should we also check for `instanceID` here? It is possible that device name is same as expected device but `describeVolume` returned stale attachment information from another instance?

jsafrane replied:
Added check for instanceID

jsafrane committed:
All check for instanceID

In case DescribeVolumes returns stale attachment and the volume was
previously attached to a different node.

gnufied commented:
/hold cancel
/lgtm
